File: qLearningAgents.py
from featureExtractors import *
import util

class QLearningAgent():
    """
      Q-Learning Agent

      Functions you should fill in:
        - computeValueFromQValues
        - computeActionFromQValues
        - getQValue
        - getAction
        - update

      Instance variables you have access to
        - self.epsilon (exploration prob)
        - self.alpha (learning rate)
        - self.discount (discount rate)

      Functions you should use
        - self.getLegalActions(state)
          which returns legal actions for a state
    """
    def __init__(self, **args):
        self.qvalues = {}

    def getQValue(self, state):
        """
          Returns Q(state,action)
          Should return 0.0 if we have never seen a state
          or the Q node value otherwise
        """
        "*** YOUR CODE HERE ***"
        if state in self.qvalues:
            return self.qvalues[state]
        else:
            return 0.0


    def computeValueFromQValues(self, state):
        """
          Returns max_action Q(state,action)
          where the max is over legal actions.  Note that if
          there are no legal actions, which is the case at the
          terminal state, you should return a value of 0.0.
        """
        "*** YOUR CODE HERE ***"
        if len(self.getLegalActions(state)) == 0:
            return 0
            
        maxValue = None
        for action in self.getLegalActions(state):
            qValue = self.getQValue(state)
            if maxValue == None or qValue > maxValue:
                maxValue = qValue
                
        return maxValue
            

    def computeActionFromQValues(self, state):
        """
          Compute the best action to take in a state.  Note that if there
          are no legal actions, which is the case at the terminal state,
          you should return None.
        """
        "*** YOUR CODE HERE ***"
        maxValue = None
        policy = None
        for action in self.getLegalActions(state):
            qValue = self.getQValue(state, action)
            if maxValue == None or qValue > maxValue:
                maxValue = qValue
                policy = action
                
        return policy

    def getAction(self, state):
        """
          Compute the action to take in the current state.  With
          probability self.epsilon, we should take a random action and
          take the best policy action otherwise.  Note that if there are
          no legal actions, which is the case at the terminal state, you
          should choose None as the action.

          HINT: You might want to use util.flipCoin(prob)
          HINT: To pick randomly from a list, use random.choice(list)
        """
        # Pick Action
        legalActions = self.getLegalActions(state)
        action = None
        "*** YOUR CODE HERE ***"
        if util.flipCoin(self.epsilon):
            action = random.choice(legalActions)
        else:
            action = self.computeActionFromQValues(state)

        return action


    def update(self, state, action, nextState, reward):
        """
          The parent class calls this to observe a
          state = action => nextState and reward transition.
          You should do your Q-Value update here

          NOTE: You should never call this function,
          it will be called on your behalf
        """
        "*** YOUR CODE HERE ***"
        if (state, action) not in self.qvalues:
            self.qvalues[state] = 0.0
        
        nextStateValue = self.computeValueFromQValues(nextState)
        currentStateValue = self.qvalues[state]

        self.qvalues[state] = (1 - self.alpha) * currentStateValue + self.alpha * (reward + self.discount * nextStateValue)

    def getPolicy(self, state):
        return self.computeActionFromQValues(state)

    def getValue(self, state):
        return self.computeValueFromQValues(state)


class MahjongQAgent(QLearningAgent):
    "Exactly the same as QLearningAgent, but with different default parameters"

    def __init__(self, epsilon=0.05,gamma=0.8,alpha=0.2, numTraining=0, **args):
        """
        These default parameters can be changed from the pacman.py command line.
        For example, to change the exploration rate, try:
            python pacman.py -p PacmanQLearningAgent -a epsilon=0.1

        alpha    - learning rate
        epsilon  - exploration rate
        gamma    - discount factor
        numTraining - number of training episodes, i.e. no learning after these many episodes
        """
        args['epsilon'] = epsilon
        args['gamma'] = gamma
        args['alpha'] = alpha
        args['numTraining'] = numTraining
        self.index = 0  # This is always Player 1
        QLearningAgent.__init__(self, **args)

    def getAction(self, state):
        """
        Simply calls the getAction method of QLearningAgent and then
        informs parent of action for Pacman.  Do not change or remove this
        method.
        """
        action = QLearningAgent.getAction(self,state)
        self.doAction(state,action)
        return action



class ApproximateQAgent(MahjongQAgent):
    """
       ApproximateQLearningAgent

       You should only have to overwrite getQValue
       and update.  All other QLearningAgent functions
       should work as is.
    """
    def __init__(self, **args):
        self.features = util.Counter()
        self.weights = util.Counter()
        
        self.weights["length of longest dishonor suit"] = 20
        self.weights["length of second and third longest dishonor suit"] = -35
        self.weights["length of honor suit"] = -18
        self.weights["is at least 11 honors"] = 500
        self.weights["number of triplets"] = 25
        self.weights["number of duplets"] = 10
        self.weights["number of sequences"] = 25
        self.weights["number of good half sequences"] = 5
        self.weights["number of melds"] = 400
        self.weights["number of melds with eyes"] = 200
        self.weights["number of orphans"] = -1
        self.weights["is 13 orphans"] = 2500
        self.weights["is win"] = 1000
        self.weights["is 10 score"] = 6400
        self.weights["is 8 score"] = 3200
        self.weights["is 7 score"] = 2400
        self.weights["is 6 score"] = 1600
        self.weights["is 3 score"] = 400
        self.weights["remaining pool count"] = 1
        
        self.alpha = 0.05
        self.discount = 0.98

    def getWeights(self):
        return self.weights

    def getQValue(self, state): #getQValue(self, state, action)
        """
          Should return Q(state,action) = w * featureVector
          where * is the dotProduct operator
        """
        "*** YOUR CODE HERE ***"
        features = SimpleExtractor.getFeatures(SimpleExtractor, state)
        return sum([self.weights[feature] * features[feature] for feature in features])

    def update(self, state, nextState, reward): #update(self, state, action, nextState, reward):
        """
           Should update your weights based on transition
        """
        "*** YOUR CODE HERE ***"
        features = SimpleExtractor.getFeatures(SimpleExtractor, state)
        # difference = reward + self.discount * self.getValue(nextState) - self.getQValue(state)
        difference = reward + self.discount * self.getQValue(nextState) - self.getQValue(state)
        for feature in features:
            self.weights[feature] = self.weights[feature] + self.alpha * difference * features[feature]

'''
    def final(self, state):
        "Called at the end of each game."
        # call the super-class final method
        PacmanQAgent.final(self, state)

        # did we finish training?
        if self.episodesSoFar == self.numTraining:
            # you might want to print your weights here for debugging
            "*** YOUR CODE HERE ***"
            pass
'''

File: C#/Reinforcement_learning/Reinforcement_Learning/QLearningAgents.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System;

namespace Reinforcement_Learning
{
    public static class ApproximateQAgent
    {
        static Dictionary<string, float> features = new Dictionary<string, float>();
        static Dictionary<string, float> weights = new Dictionary<string, float>
            {
                { "length of longest dishonor suit", 40 },
                { "length of second longest dishonor suit", -30 },
                { "length of third longest dishonor suit", -40 },
                { "length of honor suit", -5 },
                { "is at least 11 honors", 500 },
                { "is 12 longest dishonor suit", 200 },
                { "is 13 longest dishonor suit", 400 },
                { "is 14 longest dishonor suit", 600 },
                { "is 2 second longest dishonor suit", 100 },
                { "is 1 second longest dishonor suit", 300 },
                { "is 0 second longest dishonor suit", 500 },
                { "number of prevailing wind honor", 12},
                { "number of dealer honor", 14},
                { "number of both prevailing wind and dealer honor", 30},
                { "number of dragon honor", 13 },
                { "is all single", -600 },
                { "number of triplets", 40 },
                { "number of duplets", 20 },
                { "number of remaining duplets", 30 },
                { "number of sequences", 40 },
                { "number of good half sequences", 5 },
                { "number of first neighbours within 2", 20 },
                { "number of displayed melds", 400 },
                { "number of hand melds", 400 },
                { "number of hand melds with eyes", 200 },
                { "number of orphans", -2 },
                { "number of 2 or 8", -1 },
                { "is 13 orphans", 2500 },
                { "is win", 1000 },
                { "is 10 score", 6400 },
                { "is 8 score", 3200 },
                { "is 7 score", 2400 },
                { "is 6 score", 1600 },
                { "is 3 score", 400 },
                { "number of tiles that improves 1 meld", 15 },
                { "number of tiles that improves eyes", 2 }, //from 0 to 1
                { "number of tiles that improves triplets",  15},
                { "number of tiles that wins", 25 },
                { "number of tiles that improves remaining duplets", 1 },
                { "remaining pool count", 1 }
            };

        public static Dictionary<string, float> GetWeights()
        {
            return weights;
        }

        public static float GetQValue(State state)
        {
            features = SimpleExtractor.GetFeatures(state);
            float sum = 0;
            foreach (KeyValuePair<string, float> kvp in features)
            {
                sum += weights[kvp.Key] * kvp.Value;
            }
            return sum;
        }

        public static int GetDiscardDecision(State state)
        {
            List<float> allScores = new List<float>();
            List<Tile> tiles_hand = state.ownPlayer.tiles_hand;
            State state1;
            state.Print();

            Main.DebugLog("Reinforcement_Learning: original score: " + GetQValue(state).ToString());
            foreach (Tile tile in tiles_hand)
            {
                state1 = new State();
                state1.DeepCopy(state);
                state1.ownPlayer.tiles_hand.Remove(tile);
                state1.discarded_pool.Add(tile);
                state1.UpdatePossiblePool();
                allScores.Add(GetQValue(state1));
            }

            float maxScore = -1000000;
            int maxIndex = 0;
            List<int> maxIndexes = new List<int>();
            for (int i = 0; i < allScores.Count; i++)
            {
                Main.DebugLog("Reinforcement_Learning: score after discarding " + tiles_hand[i].tile_code + " : " + allScores[i].ToString());
                if (i == 0)
                {
                    maxScore = allScores[0];
                    maxIndex = 0;
                    maxIndexes.Add(0);
                    continue;
                }
                if (allScores[i] > maxScore)
                {
                    maxScore = allScores[i];
                    maxIndex = i;
                    maxIndexes.Clear();
                    maxIndexes.Add(i);
                }
                if (allScores[i] == maxScore)
                {
                    maxIndexes.Add(i);
                }
            }
            System.Random rnd = new System.Random();
            maxIndex = maxIndexes[rnd.Next(0, maxIndexes.Count)];
            return maxIndex;
        }
    }
}
